icicle_v3/backend/cpu/include/tasks_manager.h

```diff
@@ -0,0 +1,309 @@
+}
+
+void TaskBase::dispatch() {
+  (*father_fifo_tail)++;
+  status.store(DISPATCHED, std::memory_order_release);
+}
+
+void TaskBase::set_working() {
+  status.store(WORKING, std::memory_order_release);
+}
+
+void TaskBase::set_pending_result() {
+  status.store(PENDING_RESULT, std::memory_order_release);
+}
+
+void TaskBase::set_handled_result() {
+  status.store(IDLE, std::memory_order_release);
+}
+
+void TaskBase::wait_done() {
+  while (!has_result()) {
+    std::this_thread::sleep_for(std::chrono::microseconds(MANAGER_SLEEP_USEC));
+  }
+}
+
+template<class Task>
+TasksManager<Task>::Worker::Worker() 
+: mTasksFifo(TASKS_PER_THREAD), 
+  tail(0),
+  head(0),
+  kill(false)
+  {
+    // Tail idx linking can only be done after task and worker initialization
+    for (Task& task : mTasksFifo) task.link_father_tail(&tail);
+    // Init thread only after finishing all other setup to avoid data races
+    task_executor = std::thread(&TasksManager<Task>::Worker::run, this);
+  }
+
+template<class Task>
+TasksManager<Task>::Worker::~Worker() {
+  kill = true;
+  task_executor.join();
+}
+
+template<class Task>
+void TasksManager<Task>::Worker::run() {
+  while (true) {
+    for (head = 0; head < mTasksFifo.size(); head++)
+    {      
+      Task* task = &mTasksFifo[head];
+      if (!task->is_ready_for_work()) {
+        // Sleep as the thread apparently isn't fully utilized currently
+        std::this_thread::sleep_for(std::chrono::microseconds(THREAD_SLEEP_USEC));       
+        continue;
+      }
+      task->set_working();
+      task->execute();
+      task->set_pending_result();
+    }
+    if (kill) return;
+  }
+}
+
+template<class Task>
+bool TasksManager<Task>::Worker::get_free_task(Task*& task_slot) {
+  for (int i = 0; i < mTasksFifo.size(); i++)
+  {
+    // TASKS_PER_WORKER is a power of 2 so modulo is done via bitmask.
+    int tail_adjusted_idx = (i + tail) & TASK_IDX_MASK; 
+
+    if (mTasksFifo[tail_adjusted_idx].can_push_task()) // Either IDLE or PENDING_RESULT are valid.
+    {
+      task_slot = &mTasksFifo[tail_adjusted_idx];
+      return task_slot->has_result();
+    }
+  }
+  task_slot = nullptr;
+  return false;
+}
+
+template<class Task>
+void TasksManager<Task>::Worker::get_completed_task(Task*& task_slot) {
+  for (int i = 0; i < mTasksFifo.size(); i++)
+  {
+    int tail_adjusted_idx = (i + tail) & TASK_IDX_MASK;
+
+    if (mTasksFifo[tail_adjusted_idx].has_result()) // Only PENDING_RESULT is valid.
+    {
+      task_slot = &mTasksFifo[tail_adjusted_idx];
+      return;
+    }
+  }
+  task_slot = nullptr;
+}
+
+template<class Task>
+bool TasksManager<Task>::Worker::are_all_idle()
+{
+  bool all_tasks_idle = true;
+  for (Task& task : mTasksFifo) all_tasks_idle = all_tasks_idle && task.is_idle();
+  return all_tasks_idle;
+}
+
+template<class Task>
+bool TasksManager<Task>::get_free_task(Task*& task_slot) {
+  bool has_task = false;
+  do
+  {
+    for (Worker& worker : workers)
+    {
```

Comment:
I think you should have the sleep here and then quickly scan all workers instead of sleeping there